File: python/design_pattern/command/main_command.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Command Pattern

The Command pattern encapsulates a request as an object, thereby letting you
parameterize clients with different requests, queue or log requests, and
support undoable operations.

Key Components:
- Command: Interface for executing an operation
- ConcreteCommand: Binds a receiver to an action
- Invoker: Asks the command to carry out the request
- Receiver: Knows how to perform the operations
- Client: Creates commands and sets their receivers
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime


# Command interface
class Command(ABC):
    """Abstract command interface."""

    @abstractmethod
    def execute(self) -> str:
        pass

    @abstractmethod
    def undo(self) -> str:
        pass

    @abstractmethod
    def get_description(self) -> str:
        pass


# Receivers
class TextEditor:
    """Receiver: text editor that performs actual operations."""

    def __init__(self):
        self._text = ""
        self._clipboard = ""
        self._selection_start = 0
        self._selection_end = 0

    @property
    def text(self) -> str:
        return self._text

    def insert(self, text: str, position: int) -> None:
        self._text = self._text[:position] + text + self._text[position:]

    def delete(self, start: int, end: int) -> str:
        deleted = self._text[start:end]
        self._text = self._text[:start] + self._text[end:]
        return deleted

    def replace(self, start: int, end: int, new_text: str) -> str:
        old_text = self._text[start:end]
        self._text = self._text[:start] + new_text + self._text[end:]
        return old_text

    def get_text_range(self, start: int, end: int) -> str:
        return self._text[start:end]

    def select(self, start: int, end: int) -> None:
        self._selection_start = start
        self._selection_end = end

    def get_selection(self) -> tuple[int, int]:
        return (self._selection_start, self._selection_end)

    def copy_to_clipboard(self, text: str) -> None:
        self._clipboard = text

    def get_clipboard(self) -> str:
        return self._clipboard


# Concrete Commands
class InsertCommand(Command):
    """Command to insert text."""

    def __init__(self, editor: TextEditor, text: str, position: int):
        self._editor = editor
        self._text = text
        self._position = position

    def execute(self) -> str:
        self._editor.insert(self._text, self._position)
        return f"Inserted '{self._text}' at position {self._position}"

    def undo(self) -> str:
        self._editor.delete(self._position, self._position + len(self._text))
        return f"Undid insert of '{self._text}'"

    def get_description(self) -> str:
        return f"Insert '{self._text}' at {self._position}"


class DeleteCommand(Command):
    """Command to delete text."""

    def __init__(self, editor: TextEditor, start: int, end: int):
        self._editor = editor
        self._start = start
        self._end = end
        self._deleted_text = ""

    def execute(self) -> str:
        self._deleted_text = self._editor.delete(self._start, self._end)
        return f"Deleted '{self._deleted_text}'"

    def undo(self) -> str:
        self._editor.insert(self._deleted_text, self._start)
        return f"Undid delete, restored '{self._deleted_text}'"

    def get_description(self) -> str:
        return f"Delete from {self._start} to {self._end}"


class ReplaceCommand(Command):
    """Command to replace text."""

    def __init__(self, editor: TextEditor, start: int, end: int, new_text: str):
        self._editor = editor
        self._start = start
        self._end = end
        self._new_text = new_text
        self._old_text = ""

    def execute(self) -> str:
        self._old_text = self._editor.replace(self._start, self._end, self._new_text)
        return f"Replaced '{self._old_text}' with '{self._new_text}'"

    def undo(self) -> str:
        self._editor.replace(
            self._start, self._start + len(self._new_text), self._old_text
        )
        return f"Undid replace, restored '{self._old_text}'"

    def get_description(self) -> str:
        return f"Replace at {self._start}-{self._end} with '{self._new_text}'"


class CopyCommand(Command):
    """Command to copy text to clipboard."""

    def __init__(self, editor: TextEditor, start: int, end: int):
        self._editor = editor
        self._start = start
        self._end = end
        self._previous_clipboard = ""

    def execute(self) -> str:
        self._previous_clipboard = self._editor.get_clipboard()
        text = self._editor.get_text_range(self._start, self._end)
        self._editor.copy_to_clipboard(text)
        return f"Copied '{text}' to clipboard"

    def undo(self) -> str:
        self._editor.copy_to_clipboard(self._previous_clipboard)
        return "Restored previous clipboard content"

    def get_description(self) -> str:
        return f"Copy from {self._start} to {self._end}"


class PasteCommand(Command):
    """Command to paste from clipboard."""

    def __init__(self, editor: TextEditor, position: int):
        self._editor = editor
        self._position = position
        self._pasted_text = ""

    def execute(self) -> str:
        self._pasted_text = self._editor.get_clipboard()
        self._editor.insert(self._pasted_text, self._position)
        return f"Pasted '{self._pasted_text}' at position {self._position}"

    def undo(self) -> str:
        self._editor.delete(self._position, self._position + len(self._pasted_text))
        return f"Undid paste of '{self._pasted_text}'"

    def get_description(self) -> str:
        return f"Paste at {self._position}"


# Macro Command (Composite Command)
class MacroCommand(Command):
    """Composite command that executes multiple commands."""

    def __init__(self, name: str, commands: list[Command]):
        self._name = name
        self._commands = commands

    def execute(self) -> str:
        results = []
        for cmd in self._commands:
            results.append(cmd.execute())
        return f"Macro '{self._name}' executed:\n  " + "\n  ".join(results)

    def undo(self) -> str:
        results = []
        # Undo in reverse order
        for cmd in reversed(self._commands):
            results.append(cmd.undo())
        return f"Macro '{self._name}' undone:\n  " + "\n  ".join(results)

    def get_description(self) -> str:
        return f"Macro: {self._name} ({len(self._commands)} commands)"


# Invoker
@dataclass
class HistoryEntry:
    command: Command
    timestamp: datetime
    executed: bool


class EditorInvoker:
    """Invoker that executes commands and manages history."""

    def __init__(self, editor: TextEditor):
        self._editor = editor
        self._history: list[HistoryEntry] = []
        self._redo_stack: list[HistoryEntry] = []

    def execute(self, command: Command) -> str:
        """Execute command and add to history."""
        result = command.execute()
        entry = HistoryEntry(
            command=command, timestamp=datetime.now(), executed=True
        )
        self._history.append(entry)
        self._redo_stack.clear()  # Clear redo stack on new command
        return result

    def undo(self) -> str:
        """Undo last command."""
        if not self._history:
            return "Nothing to undo"

        entry = self._history.pop()
        result = entry.command.undo()
        entry.executed = False
        self._redo_stack.append(entry)
        return result

    def redo(self) -> str:
        """Redo last undone command."""
        if not self._redo_stack:
            return "Nothing to redo"

        entry = self._redo_stack.pop()
        result = entry.command.execute()
        entry.executed = True
        self._history.append(entry)
        return result

    def get_history(self) -> list[str]:
        """Get command history descriptions."""
        return [
            f"{entry.timestamp.strftime('%H:%M:%S')} - {entry.command.get_description()}"
            for entry in self._history
        ]

    def get_current_text(self) -> str:
        """Get current editor text."""
        return self._editor.text


def main() -> None:
    print("=" * 60)
    print("Command Pattern - Text Editor Demo")
    print("=" * 60)

    # Create receiver and invoker
    editor = TextEditor()
    invoker = EditorInvoker(editor)

    # Demo 1: Basic commands
    print("\n--- Basic Commands ---")

    cmd = InsertCommand(editor, "Hello", 0)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    cmd = InsertCommand(editor, " World", 5)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    cmd = InsertCommand(editor, "!", 11)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    # Demo 2: Undo/Redo
    print("\n--- Undo/Redo ---")
    print(invoker.undo())
    print(f"Text after undo: '{invoker.get_current_text()}'")

    print(invoker.undo())
    print(f"Text after undo: '{invoker.get_current_text()}'")

    print(invoker.redo())
    print(f"Text after redo: '{invoker.get_current_text()}'")

    # Demo 3: Replace and Delete
    print("\n--- Replace and Delete ---")
    cmd = ReplaceCommand(editor, 0, 5, "Hi")
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    cmd = DeleteCommand(editor, 2, 8)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    # Demo 4: Copy and Paste
    print("\n--- Copy and Paste ---")
    cmd = InsertCommand(editor, "Python is great", 0)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    cmd = CopyCommand(editor, 0, 6)  # Copy "Python"
    print(invoker.execute(cmd))

    cmd = PasteCommand(editor, 10)
    print(invoker.execute(cmd))
    print(f"Text: '{invoker.get_current_text()}'")

    # Demo 5: Macro Command
    print("\n--- Macro Command ---")

    # Create a macro to add formatted header
    editor2 = TextEditor()
    invoker2 = EditorInvoker(editor2)

    macro_commands = [
        InsertCommand(editor2, "=====\n", 0),
        InsertCommand(editor2, "HEADER", 6),
        InsertCommand(editor2, "\n=====", 12),
    ]

    macro = MacroCommand("Add Header", macro_commands)
    print(invoker2.execute(macro))
    print(f"Text: '{invoker2.get_current_text()}'")

    print("\nUndo macro:")
    print(invoker2.undo())
    print(f"Text: '{invoker2.get_current_text()}'")

    # Demo 6: Command History
    print("\n--- Command History ---")
    print("First editor history:")
    for entry in invoker.get_history():
        print(f"  {entry}")

    # Demo 7: Multiple undos
    print("\n--- Multiple Undos ---")
    print(f"Current text: '{invoker.get_current_text()}'")
    for _ in range(3):
        result = invoker.undo()
        print(result)
        print(f"  Text: '{invoker.get_current_text()}'")

    print("\n" + "=" * 60)
    print("Benefits of Command Pattern:")
    print("=" * 60)
    print("1. Decouples invoker from receiver")
    print("2. Commands are first-class objects")
    print("3. Easy to add new commands")
    print("4. Supports undo/redo operations")
    print("5. Supports macros (composite commands)")
    print("6. Commands can be queued, logged, or serialized")


if __name__ == "__main__":
    main()

File: java/design_pattern/command/MainCommand.java
/**
 * Comprehensive demonstration of the Command Pattern in Java
 *
 * The Command pattern encapsulates a request as an object, thereby letting you
 * parameterize clients with different requests, queue or log requests, and
 * support undoable operations.
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

// Command interface
interface Command {
    void execute();
    void undo();
    String getDescription();
}

// Receiver - the actual business logic
class Light {
    private String location;
    private boolean isOn = false;
    private int brightness = 100;

    public Light(String location) {
        this.location = location;
    }

    public void on() {
        isOn = true;
        System.out.println("  [Light] " + location + " light is ON");
    }

    public void off() {
        isOn = false;
        System.out.println("  [Light] " + location + " light is OFF");
    }

    public void dim(int level) {
        brightness = level;
        System.out.println("  [Light] " + location + " light dimmed to " + level + "%");
    }

    public int getBrightness() { return brightness; }
    public boolean isOn() { return isOn; }
}

class Stereo {
    private String location;
    private boolean isOn = false;
    private int volume = 5;

    public Stereo(String location) {
        this.location = location;
    }

    public void on() {
        isOn = true;
        System.out.println("  [Stereo] " + location + " stereo is ON");
    }

    public void off() {
        isOn = false;
        System.out.println("  [Stereo] " + location + " stereo is OFF");
    }

    public void setVolume(int volume) {
        this.volume = volume;
        System.out.println("  [Stereo] " + location + " stereo volume set to " + volume);
    }

    public int getVolume() { return volume; }
}

class GarageDoor {
    public void up() { System.out.println("  [GarageDoor] Garage door is open"); }
    public void down() { System.out.println("  [GarageDoor] Garage door is closed"); }
}

// Concrete Commands
class LightOnCommand implements Command {
    private Light light;

    public LightOnCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() { light.on(); }

    @Override
    public void undo() { light.off(); }

    @Override
    public String getDescription() { return "Turn light on"; }
}

class LightOffCommand implements Command {
    private Light light;

    public LightOffCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() { light.off(); }

    @Override
    public void undo() { light.on(); }

    @Override
    public String getDescription() { return "Turn light off"; }
}

class LightDimCommand implements Command {
    private Light light;
    private int newLevel;
    private int previousLevel;

    public LightDimCommand(Light light, int level) {
        this.light = light;
        this.newLevel = level;
    }

    @Override
    public void execute() {
        previousLevel = light.getBrightness();
        light.dim(newLevel);
    }

    @Override
    public void undo() {
        light.dim(previousLevel);
    }

    @Override
    public String getDescription() { return "Dim light to " + newLevel + "%"; }
}

class StereoOnCommand implements Command {
    private Stereo stereo;

    public StereoOnCommand(Stereo stereo) {
        this.stereo = stereo;
    }

    @Override
    public void execute() {
        stereo.on();
        stereo.setVolume(5);
    }

    @Override
    public void undo() { stereo.off(); }

    @Override
    public String getDescription() { return "Turn stereo on"; }
}

class GarageDoorOpenCommand implements Command {
    private GarageDoor door;

    public GarageDoorOpenCommand(GarageDoor door) {
        this.door = door;
    }

    @Override
    public void execute() { door.up(); }

    @Override
    public void undo() { door.down(); }

    @Override
    public String getDescription() { return "Open garage door"; }
}

// Macro Command - executes multiple commands
class MacroCommand implements Command {
    private Command[] commands;
    private String name;

    public MacroCommand(String name, Command[] commands) {
        this.name = name;
        this.commands = commands;
    }

    @Override
    public void execute() {
        for (Command command : commands) {
            command.execute();
        }
    }

    @Override
    public void undo() {
        for (int i = commands.length - 1; i >= 0; i--) {
            commands[i].undo();
        }
    }

    @Override
    public String getDescription() { return name; }
}

// Invoker with undo support
class RemoteControl {
    private Command[] onCommands;
    private Command[] offCommands;
    private Stack<Command> undoStack;

    public RemoteControl() {
        onCommands = new Command[7];
        offCommands = new Command[7];
        undoStack = new Stack<>();

        Command noCommand = new Command() {
            public void execute() {}
            public void undo() {}
            public String getDescription() { return "No command"; }
        };

        for (int i = 0; i < 7; i++) {
            onCommands[i] = noCommand;
            offCommands[i] = noCommand;
        }
    }

    public void setCommand(int slot, Command onCommand, Command offCommand) {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }

    public void onButtonPressed(int slot) {
        onCommands[slot].execute();
        undoStack.push(onCommands[slot]);
    }

    public void offButtonPressed(int slot) {
        offCommands[slot].execute();
        undoStack.push(offCommands[slot]);
    }

    public void undoButtonPressed() {
        if (!undoStack.isEmpty()) {
            Command command = undoStack.pop();
            System.out.println("  Undoing: " + command.getDescription());
            command.undo();
        } else {
            System.out.println("  Nothing to undo");
        }
    }
}

// Command queue for delayed execution
class CommandQueue {
    private List<Command> queue = new ArrayList<>();

    public void addCommand(Command command) {
        queue.add(command);
        System.out.println("  Queued: " + command.getDescription());
    }

    public void executeAll() {
        System.out.println("\n  Executing queued commands:");
        for (Command command : queue) {
            command.execute();
        }
        queue.clear();
    }
}

// Text editor example with undo
class TextEditor {
    private StringBuilder content = new StringBuilder();
    private Stack<Command> history = new Stack<>();

    public void executeCommand(Command command) {
        command.execute();
        history.push(command);
    }

    public void undo() {
        if (!history.isEmpty()) {
            Command command = history.pop();
            command.undo();
        }
    }

    public void append(String text) {
        content.append(text);
    }

    public void delete(int length) {
        int start = Math.max(0, content.length() - length);
        content.delete(start, content.length());
    }

    public String getContent() {
        return content.toString();
    }
}

class AppendCommand implements Command {
    private TextEditor editor;
    private String text;

    public AppendCommand(TextEditor editor, String text) {
        this.editor = editor;
        this.text = text;
    }

    @Override
    public void execute() {
        editor.append(text);
        System.out.println("  Appended: \"" + text + "\"");
    }

    @Override
    public void undo() {
        editor.delete(text.length());
        System.out.println("  Undid append: \"" + text + "\"");
    }

    @Override
    public String getDescription() { return "Append text"; }
}

public class MainCommand {
    public static void main(String[] args) {
        System.out.println("=== Command Pattern Demonstration ===\n");

        // Remote control example
        System.out.println("--- 1. Smart Home Remote Control ---");

        RemoteControl remote = new RemoteControl();

        Light livingRoomLight = new Light("Living Room");
        Light kitchenLight = new Light("Kitchen");
        Stereo stereo = new Stereo("Living Room");
        GarageDoor garageDoor = new GarageDoor();

        remote.setCommand(0, new LightOnCommand(livingRoomLight), new LightOffCommand(livingRoomLight));
        remote.setCommand(1, new LightOnCommand(kitchenLight), new LightOffCommand(kitchenLight));
        remote.setCommand(2, new StereoOnCommand(stereo), new Command() {
            public void execute() { stereo.off(); }
            public void undo() { stereo.on(); }
            public String getDescription() { return "Turn stereo off"; }
        });
        remote.setCommand(3, new GarageDoorOpenCommand(garageDoor), new Command() {
            public void execute() { garageDoor.down(); }
            public void undo() { garageDoor.up(); }
            public String getDescription() { return "Close garage door"; }
        });

        System.out.println("\nPressing buttons:");
        remote.onButtonPressed(0);   // Living room light on
        remote.onButtonPressed(1);   // Kitchen light on
        remote.onButtonPressed(2);   // Stereo on

        System.out.println("\nUndoing last command:");
        remote.undoButtonPressed();

        System.out.println("\nMore actions:");
        remote.offButtonPressed(0);  // Living room light off
        remote.undoButtonPressed();  // Undo: light back on
        System.out.println();

        // Macro command example
        System.out.println("--- 2. Macro Command (Party Mode) ---");

        Light partyLight1 = new Light("Party Room 1");
        Light partyLight2 = new Light("Party Room 2");
        Stereo partyStereo = new Stereo("Party Room");

        Command[] partyOn = {
            new LightOnCommand(partyLight1),
            new LightOnCommand(partyLight2),
            new LightDimCommand(partyLight1, 30),
            new LightDimCommand(partyLight2, 30),
            new StereoOnCommand(partyStereo)
        };

        MacroCommand partyMacro = new MacroCommand("Party Mode", partyOn);

        System.out.println("\nActivating Party Mode:");
        partyMacro.execute();

        System.out.println("\nDeactivating Party Mode (undo):");
        partyMacro.undo();
        System.out.println();

        // Command queue example
        System.out.println("--- 3. Command Queue (Scheduled Tasks) ---");

        CommandQueue queue = new CommandQueue();
        queue.addCommand(new LightOnCommand(new Light("Bedroom")));
        queue.addCommand(new LightDimCommand(new Light("Bedroom"), 50));
        queue.addCommand(new GarageDoorOpenCommand(new GarageDoor()));

        queue.executeAll();
        System.out.println();

        // Text editor with undo
        System.out.println("--- 4. Text Editor with Undo ---");

        TextEditor editor = new TextEditor();

        editor.executeCommand(new AppendCommand(editor, "Hello "));
        editor.executeCommand(new AppendCommand(editor, "World"));
        editor.executeCommand(new AppendCommand(editor, "!"));

        System.out.println("  Content: \"" + editor.getContent() + "\"");

        System.out.println("\nUndoing last two operations:");
        editor.undo();
        editor.undo();

        System.out.println("  Content: \"" + editor.getContent() + "\"");

        System.out.println("\n=== Summary ===");
        System.out.println("Command pattern benefits:");
        System.out.println("  - Decouples invoker from receiver");
        System.out.println("  - Commands can be queued, logged, or scheduled");
        System.out.println("  - Supports undo/redo operations");
        System.out.println("  - Can compose commands into macros");
    }
}
